src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,217 @@
+  shortDeclName   : Name
+  declName        : Name
+  levelNames      : List Name
+  binders         : TSyntaxArray ``Lean.Parser.Term.bracketedBinder
+  type            : Term
+  ctors           : Array CoInductiveView.CtorView
+  /- derivingClasses : Array Lean.Elab.DerivingClassView -/
+  /- computedFields  : Array Lean.Elab.Command.ComputedFieldView -/
+  deriving Inhabited
+
+namespace CoInductiveView
+
+def CtorView.ofStx (declName : Name) (modifiers : Modifiers) (ref : Syntax) : CommandElabM CoInductiveView.CtorView := do
+  let mut ctorModifiers ← elabModifiers ref[2]
+  if let some leadingDocComment := ref[0].getOptional? then
+    if ctorModifiers.docString?.isSome then
+      logErrorAt leadingDocComment "duplicate doc string"
+    ctorModifiers := { ctorModifiers with docString? := TSyntax.getDocString ⟨leadingDocComment⟩ }
+  if ctorModifiers.isPrivate && modifiers.isPrivate then
+    throwError "invalid 'private' constructor in a 'private' inductive datatype"
+  if ctorModifiers.isProtected && modifiers.isPrivate then
+    throwError "invalid 'protected' constructor in a 'private' inductive datatype"
+
+  checkValidCtorModifier ctorModifiers
+  let ctorName := ref.getIdAt 3
+  let ctorName := declName ++ ctorName
+  let ctorName ← withRef ref[3] $ applyVisibility ctorModifiers.visibility ctorName
+  let (binders, type?) := expandOptDeclSig ref[4]
+  addDocString' ctorName ctorModifiers.docString?
+  addAuxDeclarationRanges ctorName ref ref[3]
+
+  return { ref, modifiers := ctorModifiers, declName := ctorName, binders := binders.getArgs.map (⟨·⟩), type? := type?.map (⟨·⟩) }
+
+def ofModifiersAndStx (modifiers : Modifiers) (decl : Syntax) : CommandElabM CoInductiveView := do
+  let (binders, type) := expandDeclSig decl[2]!
+
+  let binders := binders.getArgs.map (⟨·⟩)
+
+  let declId  := ⟨decl[1]⟩
+  let ⟨shortDeclName, declName, levelNames⟩ ← expandDeclId declId.raw modifiers
+
+  let ctors ← decl[4].getArgs.mapM $ CtorView.ofStx declName modifiers
+
+  addDeclarationRanges declName decl
+
+  return {
+    ref := ⟨decl⟩
+
+    declName
+    shortDeclName
+
+    levelNames
+    type := ⟨type⟩
+
+    binders
+    ctors
+
+    declId
+    modifiers
+  }
+
+def ofStx (stx : Syntax) := elabModifiers stx[0] >>= (ofModifiersAndStx · stx[1])
+
+def extractIds : TSyntax ``Parser.Term.bracketedBinder → Array Ident
+  | ⟨.node _ _ #[.atom _ _, .node _ `null ids, _, _,  .atom _ _]⟩ => ids.map (⟨·⟩)
+  | _ => #[]
+
+def toBinderIds (c : CoInductiveView) : Array Ident := (c.binders.map extractIds).flatten
+
+def toRelType (c : CoInductiveView) : CommandElabM Term :=
+  c.binders.reverse.foldlM (fun curr acc => `($acc:bracketedBinder → $curr)) c.type
+
+end CoInductiveView
+
+open Parser.Term in
+section
+abbrev bb := bracketedBinder
+/- Since `bb` is an alias for `bracketedBinder`, we can safely coerce syntax of these categories  -/
+instance : Coe (TSyntax ``bb) (TSyntax ``bracketedBinder) where coe x := ⟨x.raw⟩
+end
+
```

Comment:
You might want to add a few newlines in between here. Also, add a docstring over `bb` to explain why we need it.

I don't quite remember what the other alias was called (the one where we stole this trick from), but `bb` feels like a very short name to me. We should consider making it `protected`